File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        z = nums[0]
        l, r = 0, len(nums) - 1

        min_index = 0
        min_element = z
        while l <= r:
            m = l + (r-l)//2
            
            if nums[m] == target:
                return m

            if nums[m] >= z:
                l = m+1
            else:
                r = m-1
                if nums[m] < min_element:
                    min_element = nums[m]
                    min_index = m

        if target >= z:
            l = 0
            r = max(0, min_index - 1)
        else:
            l = min_index
            r = len(nums) - 1

        while l <= r:
            m = l + (r-l)//2
            if nums[m] == target:
                return m
            elif nums[m] > target:
                r = m - 1
            else:
                l = m + 1

        return -1

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {

        //find target value
        int target = 0;
        for (int num : nums) {
            target += num;
        }
        if (target % 2 != 0) {
            return false;
        }
        target /= 2;
        
        bool res = false;
        unordered_set<int> dp;
        for (int num : nums) {
            for (int i : dp) {
                cout << i << ' ';
                dp.insert(num + i);
            }
            dp.insert(num);
            cout << '\n';

            if (find(dp.begin(), dp.end(), target) != dp.end()) {
                res = true;
            }
        }

        return res;
    }
};
